src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -407,8 +410,17 @@ private static bool IsTouchDocumentActionForDocument(CompilationAndGeneratorDriv
             public bool TryGetCompilation([NotNullWhen(true)] out Compilation? compilation)
             {
                 var state = ReadState();
-                compilation = state.FinalCompilationWithGeneratedDocuments;
-                return compilation != null;
+                if (state is FinalState finalState)
+                {
+                    compilation = finalState.FinalCompilationWithGeneratedDocuments;
+                    Contract.ThrowIfNull(compilation);
+                    return true;
+                }
+                else
+                {
+                    compilation = null;
+                    return false;
+                }
```

Comment:
If we prefer the form from the previous comment, it could be used here as well. The semantics change only in that we lose the `Contract.ThrowIfNull`, although that condition is also checked within `FinalState`.

```suggestion
                compilation = (state as FinalState)?.FinalCompilationWithGeneratedDocuments;
                return compilation != null;
```